faculty/clients/serveragent.py

```diff
@@ -0,0 +1,344 @@
+    
+    Parameters
+    ----------
+    line_number : int
+        The line number within the step log.
+    content : str
+        The content of the step log.
+    """
+
+    line_number = attrib()
+    content = attrib()
+
+
+
+@attrs
+class ServerResources(object):
+    """Information about current server resource usage.
+    
+    Parameters
+    ----------
+    milli_cpus : CpuUsage
+        current CPU utilisation.
+    memory_mb : MemoryUsage
+        current memory utilisation. 
+    """
+
+    milli_cpus = attrib()
+    memory_mb = attrib()
+
+
+@attrs
+class CpuUsage(object):
+    """Current CPU usage on a server.
+    
+    Parameters
+    ----------
+    total : int
+        total CPU resource.
+    used : int
+        currently utilised CPU resource.
+    """
+
+    total = attrib()
+    used = attrib()
+
+
+@attrs
+class MemoryUsage(object):
+    """Current Memory usage on a server.
+    
+    Parameters
+    ----------
+    total : float
+        total memory resource.
+    used : float
+        current utilised memory resource.
+    cache : float
+        current cache memory usage.
+    rss: float
+        resident set size allocation. 
+    """
+
+    total = attrib()
+    used = attrib()
+    cache = attrib()
+    rss = attrib()
+
+
+class ServerAgentClient(BaseClient):
+    """Client for the Faculty server events service.
+
+    Build this client with a session directly.
+
+    >>> client = ServerAgenClient(url, session)
+
+    Parameters
+    ----------
+    url : str
+        The URL of the object storage service.
+    session : faculty.session.Session
+        The session to use to make requests.
+    """
+    def latest_environment_execution(self):
+        """Get the latest environment execution on the server.
+        
+        Returns
+        -------
+        Execution
+            The latest environment execution on a server. 
+        """
+        return self._get("/execution/latest", _ExecutionSchema())
+
+    def stream_server_events(self, endpoint):
+        """Read the server events stream from an endpoint.
+        
+        Parameters
+        ----------
+        endpoint : str
+            HTTP request endpoint.
+
+        Yields
+        ------
+        ServerSentEventMessage
+        """
+        with self._stream(endpoint) as stream:
+            for message in stream:
+                yield message
+
+    def stream_server_resources(self):
+        """Stream the resources used by the server.
+        
+        Yields
+        ------
+        ServerResources
+        """
+
+        schema = _ServerResourcesSchema()
+        for message in self.stream_server_events("/events"):
+            if message.event == SERVER_RESOURCES_EVENT:
+                yield schema.load(json.loads(message.data))
+
+    def stream_environment_execution_step_logs(self, execution_id, step_id):
+        """Read from the environment step logs.
+        
+        Parameters
+        ----------
+        execution_id : Execution.id
```

Comment:
This is semantically valid, but the actual type of this is:
```suggestion
        execution_id : uuid.UUID
```